File: sdk/python/examples/dump.py
#!/usr/bin/env python3
"""Minimal candump-equivalent, built on candetect -- works with whichever
supported adapter is attached (gs_usb-protocol, or the GCAN/ECAN/Rexon
USBCANI-V503), no code change needed either way.

usage: dump.py [bitrate]

bitrate defaults to 500000, which both backends can do. Passing anything
else only works if a gsusb-protocol adapter is what gets opened -- the
gcan backend only has a confirmed-working init sequence for 500000 (see
driver/gcan_native/README.md) and raises for anything else.
"""
import os
import sys

# Makes this script runnable standalone (python3 examples/dump.py) without
# needing PYTHONPATH set -- Python only puts the script's own directory on
# sys.path, not its parent, so the sibling ../candetect package needs this.
sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), ".."))

from candetect import NoSupportedAdapterError, open_bus


def main():
    bitrate = int(sys.argv[1]) if len(sys.argv) > 1 else 500000

    try:
        bus = open_bus()
    except NoSupportedAdapterError as e:
        print(f"error: {e}", file=sys.stderr)
        sys.exit(1)

    print(f"opened {type(bus).__module__}.{type(bus).__qualname__}", file=sys.stderr)
    with bus:
        bus.configure(bitrate=bitrate)
        bus.start()
        print(f"listening at {bitrate} bps, Ctrl-C to stop", file=sys.stderr)
        try:
            while True:
                frame = bus.recv(timeout_ms=1000)
                if frame is not None:
                    print(frame)
        except KeyboardInterrupt:
            pass


if __name__ == "__main__":
    main()

File: sdk/lua/gsusb/examples/dump.lua
#!/usr/bin/env luajit
--[[ Minimal candump-equivalent built on the Lua gsusb SDK.

Structural difference from the Python version: Python's Bus runs a
background listener thread, so its main loop just polls a queue that
fills itself. LuaJIT here has no threads at all (see bus.lua's module
doc), so this loop's own bus:recv() call *is* what reads the device --
there's no thread filling anything in the background.

Also unlike the Python version: Lua's standard library has no SIGINT
handling (no FFI here to reach one, see lua_gsusb.c's module doc), so
there's no graceful "Ctrl-C to stop" cleanup -- the process just
terminates like any other Lua script. bus:close() below only runs on a
normal exit path, not on SIGINT.

usage: luajit dump.lua [bitrate]
]]

-- Makes this script runnable standalone (luajit examples/dump.lua)
-- without needing LUA_PATH set -- Lua only searches package.path as
-- already configured, not the script's own directory, so the sibling
-- ../bus.lua module needs this (same reason sys.path.insert(0, ...)
-- exists in the Python examples).
local script_dir = (arg[0] or ""):match("^(.*/)") or "./"
package.path = script_dir .. "../?.lua;" .. package.path

local Bus = require("bus")

local function format_frame(f)
    local tags = {}
    if f.extended then table.insert(tags, "EFF") end
    if f.rtr then table.insert(tags, "RTR") end
    if f.err then table.insert(tags, "ERR") end
    local tag = #tags > 0 and (" [" .. table.concat(tags, ",") .. "]") or ""

    local hex = {}
    for i = 1, #f.data do
        table.insert(hex, string.format("%02X", f.data:byte(i)))
    end

    return string.format("Frame(id=0x%X%s, len=%d, data=%s)", f.id, tag, #f.data, table.concat(hex, " "))
end

local bitrate = tonumber(arg[1]) or 500000

local bus = Bus.new()
bus:open()
bus:configure(bitrate)
bus:start()
io.stderr:write(string.format("listening at %d bps\n", bitrate))

while true do
    local frame, err = bus:recv(1000)
    if frame then
        print(format_frame(frame))
    elseif err then
        io.stderr:write("recv failed: " .. tostring(err) .. "\n")
        break
    end
    -- err == nil, frame == nil: plain timeout, loop again
end

bus:close()
